main.md

```diff
@@ -828,7 +830,112 @@ receipt of a `WWW-Authenticate: Bearer` challenge from a protected resource
+a nonce value to include in the DPoP proof:
+!---
+```
+ HTTP/1.1 401 Unauthorized
+ WWW-Authenticate: DPoP nonce="eyJ7S_zG.eyJH0-Z.HX4w-7v",
+   error="use_nonce", error_description="AS requires nonce in DPoP proof"
+```
+!---
+Figure: HTTP 401 Response to a Token Request Without a Nonce
+
+Other fields MAY also be included in the WWW-Authenticate response.
+
+Upon receiving the nonce, the client is expected to retry its token request
+using a DPoP proof including the supplied nonce value in a `nonce` claim
+in the DPoP proof.
+An example unencoded JWT Payload of such a DPoP proof including a nonce is:
+!---
+```
+{
+  "jti": "-BwC3ESc6acc2lTc",
+  "htm": "POST",
+  "htu": "https://server.example.com/token",
+  "iat": 1562262616,
+  "nonce": "eyJ7S_zG.eyJH0-Z.HX4w-7v"
+}
+```
+!---
+Figure: DPoP Proof Payload Including a Nonce Value
+
+The nonce syntax in ABNF as used by [@RFC6749]
+(which is the same as the `scope-token` syntax) is:
+
+!---
+```
+  nonce = 1*NQCHAR
+```
+!---
+Figure: Nonce ABNF
+
+The nonce is opaque to the client.
+
+If the `nonce` claim in the DPoP proof of a token request
+does not exactly match the nonce supplied by the authorization server to the client,
+the authorization server MUST reject the request.
+The rejection response MAY include a WWW-Authenticate value
+providing a new nonce value to use for subsequent requests.
+
+## Providing a New Nonce Value
+
+It is up to the authorization server when to supply a new nonce value
+for the client to use.
+The client is expected to use the existing supplied nonce in DPoP proofs
+until the server supplies a new nonce value.
+
+The authorization server MAY supply the new once in the same way that
+the initial one was supplied:
+by using a WWW-Authenticate response as specified above.
+Of course, each time this happens it requires an extra round trip.
+
+A more efficient manner of supplying a new once value is also defined --
+by including the `nextnonce` parameter of the `Authentication-Info` header
+in the `200 OK` response from the previous request,
+as defined in Section 3.5 of [@RFC7616].
+
+The client MUST use the new nonce value supplied for the next token request,
+and for all subsequent token requests until the authorization server
+supplies a new nonce.
+
+An example 200 OK response providing a new nonce value is:
+!---
+```
+ HTTP/1.1 200 OK
+ Authentication-Info: nextnonce="eyJ7S_zG.eyJbYu3.xQmBj-1"
+```
+!---
+Figure: HTTP 200 Response Providing the Next Nonce Value
+
+# Resource Server-Provided Nonce {#RSNonce}
+
+Resource servers can also choose to provide a nonce value to be included
+in DPoP proofs sent to them.
+They do so in exactly the same way that authorization servers do.
+
+Note that the nonces provided by the two kinds of servers are different
+and MUST not be confused with one another.
+In particular, a nonce provided to the client by a particular server
+MUST only be used with that server and no other.
```

Comment:
It would be good to clarify what a server is. I would expect that a nonce must never be sent to a different origin.

Reply:
Agreed. It's maybe delineated by origin and, if one was provided, realm. Maybe?

Reply:
I'm really reluctant to add the use of `realm`, as it's a little-used feature and would create additional error conditions to check for.  For instance, what do you do if the realm is wrong?

In practice, no harm would come from using the wrong nonce at a server, other than that it wouldn't be accepted.  That's not a security problem - it's just a code bug.  And it would be quickly corrected by the developer so that the resulting code would work.

Reply:
Understand the reluctance around `realm`. I was kinda just thinking out loud. And it seemed like a proper way to further scope where a nonce would be used. But `realm` is little-used and not well understood so might be more trouble than it's worth.

Reply:
Removed uses of `realm` from the examples, in consultation with @bc-pi , as they added no value.